File: src/main/java/com/pauljoda/modularsystems/storage/tiles/TileEntityStorageExpansion.java
package com.pauljoda.modularsystems.storage.tiles;

import com.teambr.bookshelf.collections.Location;
import com.teambr.bookshelf.common.tiles.BaseTile;
import com.teambr.bookshelf.common.tiles.IOpensGui;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Modular-Systems
 * Created by Paul Davis on 7/24/2015
 */
public abstract class TileEntityStorageExpansion extends BaseTile implements IInventory, IOpensGui {

    protected Location core;
    protected List<Location> children;

    public TileEntityStorageExpansion() {
        children = new ArrayList<>();
        core = new Location();
    }

    /**
     * Called after this has been added to a network
     */
    public abstract void addedToNetwork();

    /**
     * Called right before this is removed from a network
     */
    public abstract void removedFromNetwork();

    /**
     * Called when this block is removed from the network
     */
    public void removeFromNetwork(boolean deleteSelf) {
        if(getCore() != null && deleteSelf)
            getCore().deleteFromNetwork(this);
        for(Location child : children) {
            if(worldObj.getTileEntity(child.x, child.y, child.z) != null) {
                if(worldObj.getTileEntity(child.x, child.y, child.z) instanceof TileEntityStorageExpansion) { //Just to be safe
                    ((TileEntityStorageExpansion) worldObj.getTileEntity(child.x, child.y, child.z)).removeFromNetwork(true);
                }
            }
        }
        removedFromNetwork();
        core = new Location();
        children.clear();
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
    }

    /**
     * Used to add a child to this node
     * @param childLoc The child to add to this
     */
    public void addChild(Location childLoc) {
        children.add(childLoc);
    }

    /**
     * Used to get what network this is connected to
     * @return The network we are in
     */
    public TileStorageCore getCore() {
        return worldObj.getTileEntity(core.x, core.y, core.z) instanceof TileStorageCore ? (TileStorageCore) worldObj.getTileEntity(core.x, core.y, core.z) : null;
    }

    /*******************************************************************************************************************
     *********************************************** Tile Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    @SuppressWarnings("unchecked")
    public void updateEntity() {
        if(!worldObj.isRemote && !core.isValid() && worldObj.rand.nextInt(80) == 0) { //We don't have a network, lets look for one
           searchAndConnect();
        } else if(!worldObj.isRemote && getCore() == null && new Random().nextInt(20) == 0)
            removeFromNetwork(true);
    }

    public void searchAndConnect() {
        for(ForgeDirection dir : ForgeDirection.VALID_DIRECTIONS) {
            if(getTileInDirection(dir) != null) { //Found a tile entity
                if(getTileInDirection(dir) instanceof TileStorageCore) { //Is it the core?
                    core = ((TileStorageCore)getTileInDirection(dir)).getLocation(); //Set the core location to the core
                    ((TileStorageCore)getTileInDirection(dir)).getNetwork().addNode(this.getLocation()); //Let the network know we are here
                    addedToNetwork(); //Do extras
                    return; //No need to continue
                } else if(getTileInDirection(dir) instanceof TileEntityStorageExpansion) { //Is it another expansion?
                    if(((TileEntityStorageExpansion) getTileInDirection(dir)).getCore() != null) { //Make sure the other one is in a network
                        core = ((TileEntityStorageExpansion)getTileInDirection(dir)).getCore().getLocation(); //Get the core location
                        ((TileEntityStorageExpansion)getTileInDirection(dir)).addChild(getLocation()); //Add to the other
                        addedToNetwork(); //Do extras
                        return; //We are done here
                    }
                }
            }
        }
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
    }

    @Override
    public void readFromNBT (NBTTagCompound tag) {
        super.readFromNBT(tag);
        if(tag.hasKey("ChildSize")) {
            children = new ArrayList<>();
            for(int i = 0; i < tag.getInteger("ChildSize"); i++) {
                Location childLoc = new Location();
                childLoc.readFromNBT(tag, "child" + i);
                children.add(childLoc);
            }
        }
        if(tag.hasKey("IsInNetwork")) {
            core.readFromNBT(tag);
        }
    }

    @Override
    public void writeToNBT (NBTTagCompound tag) {
        super.writeToNBT(tag);
        if(children != null && !children.isEmpty()) {
            tag.setInteger("ChildSize", children.size());
            for(int i = 0; i < children.size(); i++)
                children.get(i).writeToNBT(tag, "child" + i);
        }
        if(core.isValid() && worldObj.getTileEntity(core.x, core.y, core.z) instanceof TileStorageCore) {
            tag.setBoolean("IsInNetwork", true);
            core.writeToNBT(tag);
        }
    }

    /*******************************************************************************************************************
     ***************************************** IInventory Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public int getSizeInventory() {
        return getCore() != null ? getCore().getSizeInventory() : 0;
    }

    @Override
    public ItemStack getStackInSlot(int slot) {
        return getCore() != null ? getCore().getStackInSlot(slot) : null;
    }

    @Override
    public ItemStack decrStackSize(int slot, int amount) {
        return getCore() != null ? getCore().decrStackSize(slot, amount) : null;
    }

    @Override
    public ItemStack getStackInSlotOnClosing(int slot) {
        return getCore() != null ? getCore().getStackInSlotOnClosing(slot) : null;
    }

    @Override
    public void setInventorySlotContents(int slot, ItemStack stack) {
        if(getCore() != null) getCore().setInventorySlotContents(slot, stack);
    }

    /**
     * Returns the name of the inventory
     */
    public String getInventoryName() {
        return getCore() != null ? getCore().getInventoryName() : "";
    }

    /**
     * Returns if the inventory is named
     */
    public boolean hasCustomInventoryName() {
        return getCore() != null && getCore().hasCustomInventoryName();
    }

    @Override
    public int getInventoryStackLimit() {
        return getCore() != null ? getCore().getInventoryStackLimit() : 0;
    }

    /**
     * Do not give this method the name canInteractWith because it clashes with Container
     */
    public boolean isUseableByPlayer(EntityPlayer player) {
        return getCore() != null;
    }

    @Override
    public void openInventory() {
        if(getCore() != null) getCore().openInventory();
    }

    @Override
    public void closeInventory() {
        if(getCore() != null) getCore().closeInventory();
    }

    @Override
    public boolean isItemValidForSlot(int slot, ItemStack stack) {
        return getCore() != null && getCore().isItemValidForSlot(slot, stack);
    }

    /*******************************************************************************************************************
     ****************************************** IOpensGui Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public Object getServerGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return getCore() != null ? getCore().getServerGuiElement(ID, player, world, x, y, z) : null;
    }

    @Override
    public Object getClientGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z) {
        return getCore() != null ? getCore().getClientGuiElement(ID, player, world, x, y, z) : null;
    }
}

File: src/main/scala/com/teambr/modularsystems/storage/tiles/TileEntityStorageExpansion.scala
package com.teambr.modularsystems.storage.tiles

import com.teambr.bookshelf.common.tiles.traits.UpdatingTile
import com.teambr.modularsystems.storage.blocks.BlockStorageExpansion
import net.minecraft.block.properties.PropertyBool
import net.minecraft.client.Minecraft
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.{BlockPos, EnumFacing}
import net.minecraft.world.World

import scala.collection.mutable.ListBuffer
import scala.util.Random

/**
 * Modular-Systems
 * Created by Dyonovan on 04/08/15
 */
abstract class TileEntityStorageExpansion extends TileEntity with UpdatingTile {

    protected[storage] var core: Option[BlockPos] = None
    protected[storage] var children = new ListBuffer[BlockPos]

    /**
     * Called after this has been added to a network
     */
    def addedToNetwork(): Unit = {
        Minecraft.getMinecraft.renderGlobal.markBlockRangeForRenderUpdate(pos.getX, pos.getY, pos.getZ, pos.getX, pos.getY, pos.getZ)
    }

    /**
     * Called right before this is removed from a network
     */
    def removedFromNetwork(): Unit = {
        Minecraft.getMinecraft.renderGlobal.markBlockRangeForRenderUpdate(pos.getX, pos.getY, pos.getZ, pos.getX, pos.getY, pos.getZ)
    }

    /**
     * Called when this block is removed from the network
     */
    def removeFromNetwork(deleteSelf: Boolean): Unit = {
        if (getCore.isDefined && deleteSelf) getCore.get.deleteFromNetwork(this)

        for (child <- children) {
                worldObj.getTileEntity(child) match {
                    case expansion: TileEntityStorageExpansion => expansion.removeFromNetwork(true)
                    case _ =>
                }
        }
        removedFromNetwork()
        core = None
        children.clear()
        if (worldObj.getTileEntity(pos).isInstanceOf[TileEntityStorageExpansion])
            Minecraft.getMinecraft.renderGlobal.markBlockRangeForRenderUpdate(pos.getX, pos.getY, pos.getZ, pos.getX, pos.getY, pos.getZ)
        worldObj.markBlockForUpdate(getPos)
    }

    /**
     * Used to add a child to this node
     * @param childLoc The child to add to this
     */
    def addChild(childLoc: BlockPos) {
        if (core.isDefined) {
            getCore.get.network.addNode(childLoc)
            children += childLoc
        }
    }

    /**
     * Used to get what network this is connected to
     * @return The network we are in
     */
    def getCore: Option[TileStorageCore] = {
        if (core.isDefined && worldObj.getTileEntity(core.get).isInstanceOf[TileStorageCore])
            Some(worldObj.getTileEntity(core.get).asInstanceOf[TileStorageCore])
        else None
    }

    /*******************************************************************************************************************
      *********************************************** Tile Methods *****************************************************
      ******************************************************************************************************************/

    override def onServerTick() : Unit = {
        if (core.isEmpty && worldObj.rand.nextInt(80) == 0)
            searchAndConnect()
        else if (getCore.isEmpty && new Random().nextInt(20) == 0)
            removeFromNetwork(true)
    }

    override def onClientTick(): Unit = { }

    def searchAndConnect(): Unit = {
        for (dir <- EnumFacing.values()) {
            if (getTileInDirection(pos.add(dir.getDirectionVec)).isDefined) {
                    getTileInDirection(pos.add(dir.getDirectionVec)).get match {
                    case tile: TileStorageCore =>
                        core = Some(tile.getPos)
                        tile.getNetwork.addNode(getPos)
                        addedToNetwork()
                    case tile: TileEntityStorageExpansion =>
                        if (tile.getCore.isDefined) {
                            core = Some(tile.getCore.get.getPos)
                            tile.addChild(getPos)
                            addedToNetwork()
                        }
                    case _ =>
                }
            }
        }
        worldObj.markBlockForUpdate(getPos)
    }

    def getTileInDirection(pos: BlockPos): Option[(TileEntity)] = {
        worldObj.getTileEntity(pos) != null match {
            case true => Some(worldObj.getTileEntity(pos))
            case _ => None
        }
    }

    override def readFromNBT(tag: NBTTagCompound): Unit = {
        super[TileEntity].readFromNBT(tag)
        children = new ListBuffer[BlockPos]
        if (tag.hasKey("ChildSize")) {
            for (i <- 0 until tag.getInteger("ChildSize"))
                children += BlockPos.fromLong(tag.getLong("Child" + i))
        }
        if (tag.hasKey("IsInNetwork"))
            core = Some(BlockPos.fromLong(tag.getLong("IsInNetwork")))
        else core = None
    }

    override def writeToNBT(tag: NBTTagCompound) {
        super[TileEntity].writeToNBT(tag)
        if (children.nonEmpty) {
            tag.setInteger("ChildSize", children.size)
            for (i <- children.indices)
                tag.setLong("Child" + i, children(i).toLong)
        }
        if (core.isDefined) {
            tag.setLong("IsInNetwork", core.get.toLong)
        }
    }
}
